### PR TITLE
Dac interface

### PR DESCRIPTION
**Pull request by zx100x100**
this code sets the DAC voltages, then it will turn on the red LED if there is a problem, then blinks the blue LED indefinitely.

### CODE REVIEW

File: main.c [hunk 1]
```diff
@@ -26,6 +28,15 @@ static void LED_init() {
     TRISC7 = 0;
     LATC7 = 1;      // turn the led off
 }
+const uint8_t highVal = 200;
```
Comment (amihaila) [inline main.c]: What units are these in?

File: main.c [hunk 2]
```diff
@@ -26,6 +28,15 @@ static void LED_init() {
...
+const uint8_t lowVal = 50;
+uint8_t byte1 = 0x00, byte2 = 0x00;
```
Comment (amihaila) [inline main.c]: These don't need to be global. They should be contained within the DAC set voltage function.

File: main.c [hunk 3]
```diff
@@ -26,6 +28,15 @@ static void LED_init() {
...
+uint8_t* b1 = &byte1, b2 = &byte2;
```
Comment (amihaila) [inline main.c]: These will eventually disappear (see above comment), but as a general thing you don't need to declare the pointers here, you can just pass in `&byte1` and `&byte2` directly to your analog value to data function.

File: main.c [hunk 4]
```diff
@@ -26,6 +28,15 @@ static void LED_init() {
...
+
+void analogValToData(uint8_t value, uint8_t* B1, uint8_t* B2){
```
Comment (amihaila) [inline main.c]: A better way to do this would be to pass in a pointer to a two byte buffer instead of two separate byte pointers. The pointer to that buffer can then be passed directly to the i2c_writeNbytes function provided by the i2c driver (instead of creating a new function for writing 2 bytes).

File: main.c [hunk 5]
```diff
@@ -37,14 +48,16 @@ int main(int argc, char** argv) {
     
     LED_init();
     
+    // Set DACs
+    analogValToData(highVal, b1, b2);
```
Comment (amihaila) [inline main.c]: I think this would be neater if the analog/data conversion and I2C communication was handled in one go, i.e. something like `dac_set_voltage_i2c(uint8_t i2c_addr, <whatever type> voltage)`. That could then return the error status.

File: main.c [hunk 6]
```diff
@@ -37,14 +48,16 @@ int main(int argc, char** argv) {
...
+    i2c1_write2Bytes(HIGH_DAC_ADDRESS, byte1, byte2);
+    analogValToData(lowVal, b1, b2);
+    i2c1_write2Bytes(LOW_DAC_ADDRESS, byte1, byte2); 
+    if(i2c1_getLastError() == I2C1_FAIL_TIMEOUT)
```
Comment (amihaila) [inline main.c]: nitpick: braces around single line if-statements are preferred. This lowers the likelihood of someone in the future adding another line there while forgetting to add brackets. That type of error is a pain to debug.

File: mcc_generated_files/i2c1.h [hunk 7]
```diff
@@ -210,6 +210,9 @@ void i2c1_write2ByteRegister(uint8_t address, uint8_t reg, uint16_t data);
 */
 void i2c1_writeNBytes(uint8_t address, void * data, uint8_t len);
 
+
```
Comment (amihaila) [inline mcc_generated_files/i2c1.h:213]: Same as above, this code isn't needed. Typically you don't want to modify existing library code unless it really can't do the thing you need (or has a bug).

File: mcc_generated_files/i2c1.c [hunk 8]
```diff
@@ -205,7 +205,16 @@ void i2c1_writeNBytes(uint8_t address, void* data, uint8_t len)
     }
     wait4Stop();
 }
-
+void i2c1_write2Bytes(uint8_t address, uint8_t byte1, uint8_t byte2)
```
Comment (amihaila) [inline mcc_generated_files/i2c1.c:208]: You don't need this code. The function above (i2c_writeNbytes) already does what you need in the general case.

**Review by amihaila — CHANGES_REQUESTED** [review]
The functionality looks good. Keeping track of the I2C error status as part of the demo is a nice touch. I have a few structural comments which are detailed more in the file itself. The two main things are:
1. the DAC specific code (analog/data conversion, i2c sends) can be encapsulated more
2. you're handling the data as two discrete bytes, which is what led you to need a new i2c driver function. Instead of doing this, you can allocate a two byte buffer and then use the existing i2c library functions. This would also make things cleaner since the two data bytes form a cohesive whole, so there isn't any need to treat them separately.

An alternative way to structure this would be:
void dac_set_voltage_i2c(address, voltage) {
    uint8_t data_buf[2];
    data_buf[0] = // whatever conversion you currently do
    data_buf[1] = // whatever conversion you currently do
    i2c_writeNbytes(address, data_buf, 2);
}

**Comment by amihaila** [comment]
Also, please rebase your changes onto master when needed instead of merging (that's a general thing that makes your branch history easier to read).

File: main.c [hunk 9]
```diff
@@ -15,11 +15,18 @@
 #define WHITE_LED_OFF (LATC6 = 1)
 #define BLUE_LED_ON (LATC7 = 0)
 #define BLUE_LED_OFF (LATC7 = 1)
+#define GET_BATTERY_VOLTAGE ADCC_GetSingleConversion(RA0)
```
Comment (amihaila) [inline main.c]: Hiding a full blown function inside a macro is typically considered bad form. The code should not hide the fact that an ADC conversion is being performed. If you want to make which ADC channel is being used more clear (which I think was your intention), you should add to the `adcc_channel_enum` in the ADC header file:
```
typedef enum
{
    channel_VBAT =  0x0,
    channel_VSENSE =  0x1,
    channel_VSS =  0x3B,
    channel_Temp =  0x3C,
    channel_DAC1 =  0x3D,
    channel_FVR_Buffer1 =  0x3E,
    channel_FVR_Buffer2 =  0x3F
} adcc_channel_t;
```
MPLAB code config will do this for you if you specify the pins, otherwise you can edit manually (just be consistent).

File: main.c [hunk 10]
```diff
@@ -15,11 +15,18 @@
...
+#define GET_CURRENT_DRAW ADCC_GetSingleConversion(RA1)
```
Comment (amihaila) [inline main.c]: Same thing about function calls inside macros. The name on this one is also kind of misleading - you're not measuring current draw, you're measuring the amplified voltage drop across a resistor. Additional calcs need to be done to get the actual current draw.

File: main.c [hunk 11]
```diff
@@ -15,11 +15,18 @@
...
+#define GET_BATTERY_VOLTAGE ADCC_GetSingleConversion(RB5)
```
Comment (amihaila) [inline main.c]: `GET_BATTERY_VOLTAGE` has already been defined on line 18. The compiler complains about this (`main.c:20: warning: (111) redefining preprocessor macro "GET_BATTERY_VOLTAGE" (main.c: 18)`). It's a good idea to read the output (or turn on -Werror) in your compiler options so that you can catch things like this.

File: main.c [hunk 12]
```diff
@@ -15,11 +15,18 @@
...
 #define HIGH_DAC_ADDRESS 0x4d
 #define LOW_DAC_ADDRESS 0x4c
+#define HIGH_VAL 50
```
Comment (amihaila) [inline main.c]: Naming nitpick: these should specify what the low and high values are for (eg. `DAC_HIGH_VALUE`).

File: main.c [hunk 13]
```diff
@@ -28,39 +35,90 @@ static void LED_init() {
...
 
-void analogValToData(uint8_t value, uint8_t* B1, uint8_t* B2){
-   *B1 = ((value & 0xf0) >> 4); // shift the 4 MSB to the 4 LSB
-   *B2 = ((value & 0x0f) << 4); // mask the 4MSB then shift the 4 LSB to the 4MSB
+
+void analogValToData(uint8_t value, uint8_t data_buf[2]){
+   data_buf[0] = ((value & 0xf0) >> 4); // shift the 4 MSB to the 4 LSB
+   data_buf[1] = ((value & 0x0f) << 4); // mask the 4MSB then shift the 4 LSB to the 4MSB
+}
+void setDACs(int lowVal, int highVal){
```
Comment (amihaila) [inline main.c]: Arguments lowVal and highVal are never used. You're using the macros inside the function directly.

File: main.c [hunk 14]
```diff
@@ -28,39 +35,90 @@ static void LED_init() {
...
+  uint8_t data_buf[2];
```
Comment (amihaila) [inline main.c]: can you be more consistent with your indentation? one level =  4 spaces

File: main.c [hunk 15]
```diff
@@ -28,39 +35,90 @@ static void LED_init() {
...
+  analogValToData(HIGH_VAL, data_buf);
+  i2c1_writeNBytes(HIGH_DAC_ADDRESS, data_buf, 2);
+  analogValToData(LOW_VAL, data_buf);
+  i2c1_writeNBytes(LOW_DAC_ADDRESS, data_buf, 2); 
+  if(i2c1_getLastError() == I2C1_FAIL_TIMEOUT){
+        RED_LED_ON;
+  }
 }
 
+int openVent(){
```
Comment (amihaila) [inline main.c]: I know we didn't specify this anywhere but our preferred convention for naming functions is snake_case instead of camelCase. Putting acronyms in all-caps is fine.

File: main.c (hunk 11 above)
Comment (amihaila) [inline main.c]: It looks like RB5 and RA1 are never explicitly initialized for use with the ADC.

File: main.c (hunk 15 above)
Comment (amihaila) [reply main.c]: This applies everywhere in this file.

File: main.c [hunk 16]
```diff
@@ -28,39 +35,90 @@ static void LED_init() {
...
+    LinAc_init();
     
     // Set DACs
-    analogValToData(highVal, b1, b2);
-    i2c1_write2Bytes(HIGH_DAC_ADDRESS, byte1, byte2);
-    analogValToData(lowVal, b1, b2);
-    i2c1_write2Bytes(LOW_DAC_ADDRESS, byte1, byte2); 
-    if(i2c1_getLastError() == I2C1_FAIL_TIMEOUT)
-            RED_LED_ON;
+    setDACs(LOW_VAL, HIGH_VAL);
```
Comment (amihaila) [inline main.c]: Comment is now unnecessary

File: main.c [hunk 17]
```diff
@@ -15,11 +15,18 @@
...
 #define BLUE_LED_OFF (LATC7 = 1)
+#define GET_BATTERY_VOLTAGE ADCC_GetSingleConversion(RA0)
+#define GET_CURRENT_DRAW ADCC_GetSingleConversion(RA1)
+#define GET_BATTERY_VOLTAGE ADCC_GetSingleConversion(RB5)
 #define HIGH_DAC_ADDRESS 0x4d
 #define LOW_DAC_ADDRESS 0x4c
+#define HIGH_VAL 50
+#define LOW_VAL 3
+
+int BatV = 0;
```
Comment (amihaila) [inline main.c]: does this need to be a global variable?

File: main.c [hunk 18]
```diff
@@ -28,39 +35,90 @@ static void LED_init() {
     TRISC7 = 0;
     LATC7 = 1;      // turn the led off
 }
-const uint8_t highVal = 200;
-const uint8_t lowVal = 50;
-uint8_t byte1 = 0x00, byte2 = 0x00;
-uint8_t* b1 = &byte1, b2 = &byte2;
+static void LinAc_init() {
```
Comment (amihaila) [inline main.c]: Can you break out all the linear actuator related code into a separate pair of files (header + source).

**Review by amihaila — CHANGES_REQUESTED** [review]
Overall looks good. I haven't had the chance to test it out on my hardware, so I haven't looked at the linear actuator code in depth (waiting on a battery). I have some general style/good practice comments but nothing major.

File: main.c (hunk 17 above)
Comment (zx100x100) [reply main.c]: That is just so I can view it while debugging to make sure I was reading the analog signal correctly

File: main.c (hunk 12 above)
Comment (zx100x100) [reply main.c]: Makes sense

File: main.c (hunk 9 above)
Comment (zx100x100) [reply main.c]: sounds good

File: main.c (hunk 18 above)
Comment (zx100x100) [reply main.c]: sure

File: main.c (hunk 11 above)
Comment (zx100x100) [reply main.c]: So I also need to initialize the specific registers for the pins I'm using right?

Comment (amihaila) [reply main.c]: That's correct. If you assign those pins as ADC pins within MCC directly, it'll take care of it for you.

File: LINAC.c [hunk 19]
```diff
@@ -0,0 +1,33 @@
+#include <xc.h>
+#include "LINAC.h"
```
Comment (amihaila) [inline LINAC.c]: keep file names lowercase

File: LINAC.c [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+
+ void LINAC_init() {
```
Comment (amihaila) [inline LINAC.c]: As discussed earlier, "linac" means linear accelerator, a type of particle accelerator, so the file/function names should be changed. Something like `linear_act...` or `lin_actuator` or whatever would be more accurate.

File: LINAC.c [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+    
+    TRISB3 = 1;     //set LIM_HIGH as input
+    ANSELB3 = 0;    //set as digital input
+    
+    TRISB4 = 1;     //set LIM_LOW as input
+    ANSELB4 = 0;    //set as digital input
+    
+}
+ 
+int open_vent(){
```
Comment (amihaila) [inline LINAC.c]: Why does this return int? Should be void unless the return value means something (also, conventionally an "everything is normal" int return value is 0).

Comment (amihaila) [inline LINAC.c]: nitpick: the C language standard specifies that a function with no arguments should be declared as `open_vent(void)`. Refer to: https://wiki.sei.cmu.edu/confluence/display/c/DCL20-C.+Explicitly+specify+void+when+a+function+accepts+no+arguments 

xc8 also throws a warning in this case. make sure you read the output and you'll catch more things like this.

Note that this is not the case in C++.

File: LINAC.c [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+    LATB1 = 1;      //FWD High
+    LATB2 = 0;      //BWD Low
+    while(RB3)
```
Comment (amihaila) [inline LINAC.c]: This function should not block. It should just set the relevant pins and exit. This is so that the main event loop runs as quickly as possible. More seriously, if the valve gets stuck, the entire board is stuck in an infinite loop. Valve status will be checked independently of this function.

File: LINAC.c [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+    {}
+    return 1;
+    
+}
+int close_vent(){
```
Comment (amihaila) [inline LINAC.c]: same comments as above

File: LINAC.c [hunk 24]
```diff
@@ -0,0 +1,33 @@
...
+    LATB2 = 1;      //BWD High
+    LATB1 = 0;      //FWD Low
+    while(RB4)
+    {}
+    return 1;
+}
```
Comment (amihaila) [inline LINAC.c]: nitpick: missing newline at end of file

File: LINAC.h [hunk 25]
```diff
@@ -0,0 +1,21 @@
+/*
```
Comment (amihaila) [inline LINAC.h]: We usually delete the mplab-generated header. We have git blame to keep track of authorship which is more granular anyway.

File: main.c [hunk 26]
```diff
@@ -1,4 +1,4 @@
-#include <stdio.h>
+#ihnclude <stdio.h>
```
Comment (amihaila) [inline main.c]: typo. if this compiled, you can kill that entire line. Doesn't look like it's being used anyway.

File: main.c [hunk 27]
```diff
@@ -15,17 +17,15 @@
...
-#define HIGH_VAL 50
-#define LOW_VAL 3
+#define HIGH_DAC_VAL 7
+#define LOW_DAC_VAL 5
 
 int BatV = 0;
+int LINAC_POT = 0;
 
-static void LED_init() {
+void LED_init() {
```
Comment (amihaila) [inline main.c]: why did you remove the `static` qualifier?

File: main.c [hunk 28]
```diff
@@ -86,15 +68,16 @@ int main(int argc, char** argv) {
...
-    // Set DACs
-    setDACs(LOW_VAL, HIGH_VAL);
+    set_DACs();
     
-    closeVent();
+    //close_vent();
     while (1){
-        BatV = GET_BATTERY_VOLTAGE;
-        if(RB4){
+        BatV = ADCC_GetSingleConversion(channel_VBAT);
```
Comment (amihaila) [inline main.c]: In the future, add a comment or name the variables so that the units of the output are clear. These are raw adc values that just happen to correspond to mV due to the configuration of this particular ADC.

File: mcc_generated_files/pin_manager.h [hunk 29]
```diff
@@ -59,6 +59,78 @@
 #define PULL_UP_ENABLED      1
 #define PULL_UP_DISABLED     0
 
+// get/set channel_VBAT aliases
```
Comment (amihaila) [inline mcc_generated_files/pin_manager.h:62]: These are new. Where are they coming from? They should not be needed.

File: mcc_generated_files/tmr0.c [hunk 30]
```diff
@@ -0,0 +1,121 @@
+/**
```
Comment (amihaila) [inline mcc_generated_files/tmr0.c]: As discussed, this file should be deleted. I can deal with it if mcc gives too much trouble.

File: main.c (hunk 28 above)
Comment (amihaila) [inline main.c]: How is your ADC configured?

**Review by amihaila — CHANGES_REQUESTED** [review]

File: mcc_generated_files/tmr0.c (hunk 30 above)
Comment (zx100x100) [reply mcc_generated_files/tmr0.c]: I figured it out, I didn't realize I needed to clean and build after deleting it

File: mcc_generated_files/pin_manager.h (hunk 29 above)
Comment (zx100x100) [reply mcc_generated_files/pin_manager.h:62]: These where generated by the MCC, I can try to delete them if needed.

**Comment by amihaila** [comment]
Don't bother manually modifyingnth the MCC files

File: lin_actuator.h [hunk 31]
```diff
@@ -0,0 +1,36 @@
...
+    closed = 2,
+    opening = 3,
+    closing = 4,
+    stopped_between_states =  5,
+    in_motion =  6,
+    stopped = 7
+} lin_actuator_states;
+
+
+lin_actuator_states moving = stopped;
```
Comment (amihaila) [inline lin_actuator.h]: You want to put this in lin_actuator.c (and declare it static unless you have a good reason to do otherwise). The way it is now, this variable is public to all files that include lin_actuator.h which it shouldn't be.

File: lin_actuator.h [hunk 32]
```diff
@@ -0,0 +1,36 @@
...
+ * File:   LINAC.h
+ * Author: zx100
+ *
+ * Created on February 10, 2019, 12:25 AM
+ */
+
+#ifndef LINAC_H
+#define	LINAC_H
+
+typedef enum
```
Comment (amihaila) [inline lin_actuator.h:22]: note: enums in c increment automatically, you don't need to assign a value to each one.

File: main.c [hunk 33]
```diff
@@ -15,9 +17,18 @@
 #define WHITE_LED_OFF (LATC6 = 1)
 #define BLUE_LED_ON (LATC7 = 0)
 #define BLUE_LED_OFF (LATC7 = 1)
+#define HIGH_DAC_ADDRESS 0x4d
```
Comment (amihaila) [inline main.c]: These would make sense to be put in lin_actuator.h

File: main.c [hunk 34]
```diff
@@ -27,27 +38,72 @@ static void LED_init() {
     LATC7 = 1;      // turn the led off
 }
 
+void set_DACs(){
```
Comment (amihaila) [inline main.c]: It would make sense to put this in lin_actuator.c since it's directly related to that functionality.

File: main.c [hunk 35]
```diff
@@ -27,27 +38,72 @@ static void LED_init() {
...
+  i2c1_writeNBytes(HIGH_DAC_ADDRESS, data_buf, 2);
+  data_buf[0] = ((LOW_DAC_VAL & 0xf0) >> 4); // shift the 4 MSB to the 4 LSB
+  data_buf[1] = ((LOW_DAC_VAL) << 4); // mask the 4MSB then shift the 4 LSB to the 4MSB
+  i2c1_writeNBytes(LOW_DAC_ADDRESS, data_buf, 2); 
+  if(i2c1_getLastError() == I2C1_FAIL_TIMEOUT){
+        RED_LED_ON;
+  }
+}
+
+int check_battery_voltage(void){    //returns mV
```
Comment (amihaila) [inline main.c:40]: Comment for where the magic number is coming from would be good (for future readers of this code).

**Review by amihaila — APPROVED** [review]

File: lin_actuator.h (hunk 31 above)
Comment (zx100x100) [reply lin_actuator.h]: what does static actually mean?

Comment (amihaila) [reply lin_actuator.h]: https://stackoverflow.com/questions/572547/what-does-static-mean-in-c

Comment (amihaila) [reply lin_actuator.h]: in your situation, it declares the variable as local to that file

Comment (amihaila) [reply lin_actuator.h]: it's a way to declare that that variable/function should not be modified by code that isn't in that file.

File: lin_actuator.h [hunk 36]
```diff
@@ -1,27 +1,37 @@
 /*
```
Comment (amihaila) [inline lin_actuator.h:1]: This header comment should just go. We usually have multiple authors per file and git keeps track of that much better. Not a must-fix but I'll remove this once merged.

File: lin_actuator.h [hunk 37]
```diff
@@ -1,27 +1,37 @@
...
- * Author: zx100
+ * Author: Zachariah Mears
  *
  * Created on February 10, 2019, 12:25 AM
  */
 
 #ifndef LINAC_H
 #define	LINAC_H
 
+#define RED_LED_ON (LATC5 = 0)
```
Comment (amihaila) [inline lin_actuator.h:11]: LED macros defined twice. Compiler should complain about this.

Comment (zx100x100) [reply lin_actuator.h:11]: the compiler isn't complaining, although should I delete the macros in main.c?

File: lin_actuator.h (hunk 36 above)
Comment (zx100x100) [reply lin_actuator.h:1]: oops, i think mplab made it for some reason

File: lin_actuator.h (hunk 37 above)
Comment (amihaila) [reply lin_actuator.h:11]: you should delete the led macros in lin_actuator.h. they don't have anything to do with the valve.

Comment (amihaila) [reply lin_actuator.h:11]: by "complain" i mean it should issue a warning along the lines of "macro redefinition at line xx." It will not fail the build.

Comment (zx100x100) [reply lin_actuator.h:11]: ok, currently set_DACs uses them

Comment (zx100x100) [reply lin_actuator.h:11]: I can delete that line though

Comment (amihaila) [reply lin_actuator.h:11]: ok you can leave them then

**Comment by amihaila** [comment]
Squash merged manually into master.
